DllImportGenerator/DllImportGenerator/Marshalling/DelegateMarshaler.cs

```diff
@@ -0,0 +1,115 @@
+                                ConditionalExpression(
+                                    BinaryExpression(
+                                        SyntaxKind.NotEqualsExpression,
+                                        IdentifierName(managedIdentifier),
+                                        LiteralExpression(SyntaxKind.NullLiteralExpression)
+                                    ),
+                                    InvocationExpression(
+                                        ParseName("global::System.Runtime.InteropServices.Marshal.GetFunctionPointerForDelegate"),
+                                        ArgumentList(SingletonSeparatedList(Argument(IdentifierName(managedIdentifier))))),
+                                    LiteralExpression(SyntaxKind.DefaultLiteralExpression))));
+                    }
+                    break;
+                case StubCodeContext.Stage.Unmarshal:
+                    if (info.IsManagedReturnPosition || (info.IsByRef && info.RefKind != RefKind.In))
+                    {
+                        yield return ExpressionStatement(
+                            AssignmentExpression(
+                                SyntaxKind.SimpleAssignmentExpression,
+                                IdentifierName(managedIdentifier),
+                                ConditionalExpression(
+                                    BinaryExpression(
+                                        SyntaxKind.NotEqualsExpression,
+                                        IdentifierName(nativeIdentifier),
+                                        LiteralExpression(SyntaxKind.DefaultLiteralExpression)),
+                                    InvocationExpression(
+                                        MemberAccessExpression(
+                                            SyntaxKind.SimpleMemberAccessExpression,
+                                            ParseName("global::System.Runtime.InteropServices.Marshal"),
+                                            GenericName(Identifier("GetDelegateForFunctionPointer"))
+                                            .WithTypeArgumentList(
+                                                TypeArgumentList(
+                                                    SingletonSeparatedList(
+                                                        info.ManagedType.AsTypeSyntax())))),
+                                        ArgumentList(SingletonSeparatedList(Argument(IdentifierName(nativeIdentifier))))),
+                                    LiteralExpression(SyntaxKind.NullLiteralExpression))));
+                    }
+                    break;
+                case StubCodeContext.Stage.Cleanup:
+                    if (info.RefKind != RefKind.Out)
+                    {
+                        yield return ExpressionStatement(
+                            InvocationExpression(
+                                ParseName("global::System.GC.KeepAlive"),
```

Comment:
We could definitely add a "keep alive" stage or something like that. This is used in the runtime marshalling by delegates and a few of the legacy handle types (HandleRef and CriticalHandle) and runtime handles (which are only useful in QCall scenarios)

Reply:
I am fine with a "KeepAlive" stage. I only mention this because we are greenfield here and can learn from the past. I want to ensure we name things for clarity. Should we change "Clean" to "PostInvoke"? That is also an option in my opinion.

@elinor-fung Thoughts?

Reply:
I think we should keep the "Clean" name because we want to reuse the stage names for other generators that don't have an "invoke" concept such as struct marshalling, which still needs the "clean" concept (although it will also put the keep-alive in the same function as clean to simplify codegen).